agent/directory.go

```diff
@@ -57,6 +54,74 @@ func (a *Agent) createTaskDirectory(tc *taskContext, taskDir string) (string, er
 	return taskDir, nil
 }
 
+// generateTaskDirectoryName generates a task directory path. Typically, this is
+// in the format: <agent_working_directory>/<hash>.
+//
+// On Windows hosts, the hash portion is shorter than on other distros because
+// it has a max file path length. Having a long task directory's path can cause
+// issues for some tasks that have deep paths that hit the Windows path length
+// limit. Shortening the typical 32-character directory name reduces the
+// problem.
+func (a *Agent) generateTaskDirectoryName(tc *taskContext) (string, error) {
+	dirName, err := a.generateTaskDirectoryHash(fmt.Sprintf("%s_%d_%d", tc.taskConfig.Task.Id, tc.taskConfig.Task.Execution, os.Getpid()))
+	if err != nil {
+		return "", errors.Wrap(err, "generating hash for randomized task directory")
+	}
+
+	if runtime.GOOS != "windows" {
+		return filepath.Join(a.opts.WorkingDirectory, dirName), nil
```

Comment:
Not a nit but we do the same filepath.Join call below, we could make the `taskDirPath` above this if statement and reuse the variable. I think the comment staying with the `if !fileExists` makes sense